pyteal/ast/abi/address.py

```diff
@@ -1,33 +1,120 @@
+        super().__init__(AddressTypeSpec())
 
     def type_spec(self) -> AddressTypeSpec:
         return AddressTypeSpec()
 
     def get(self) -> Expr:
         return self.stored_value.load()
 
+    def set(
+        self,
+        value: Union[
+            Sequence[Byte],
+            StaticArray[Byte, Literal[32]],
+            ComputedValue[StaticArray[Byte, Literal[32]]],
+            "Address",
+            str,
+            bytes,
+            Expr,
+        ],
+    ):
+
+        match value:
+            case ComputedValue():
+                pts = value.produced_type_spec()
+                if pts == AddressTypeSpec() or pts == StaticArrayTypeSpec(
+                    ByteTypeSpec(), AddressLength.Bytes.value
+                ):
+                    return value.store_into(self)
+
+                raise TealInputError(
+                    f"Got ComputedValue with type spec {pts}, expected AddressTypeSpec or StaticArray[Byte, Literal[32]]"
+                )
+            case BaseType():
+                if (
+                    value.type_spec() == AddressTypeSpec()
+                    or value.type_spec()
+                    == StaticArrayTypeSpec(ByteTypeSpec(), AddressLength.Bytes.value)
+                ):
+                    return self.stored_value.store(value.stored_value.load())
+
+                raise TealInputError(
+                    f"Got {value} with type spec {value.type_spec()}, expected AddressTypeSpec"
+                )
+            case str():
+                if len(value) == AddressLength.String.value:
+                    return self.stored_value.store(Addr(value))
+                raise TealInputError(
+                    f"Got string with length {len(value)}, expected {AddressLength.String.value}"
+                )
+            case bytes():
+                if len(value) == AddressLength.Bytes.value:
+                    return self.stored_value.store(Bytes(value))
+                raise TealInputError(
+                    f"Got bytes with length {len(value)}, expected {AddressLength.Bytes.value}"
+                )
+            case Expr():
+                return self.stored_value.store(value)
+            case CollectionSequence():
+                # TODO: mypy thinks its possible for the type of `value` here to be
+                # Union[Sequence[Byte], str, bytes] even though we check above?
+                if isinstance(value, str) or isinstance(value, bytes):
+                    return
+                return super().set(value)
+
+        # TODO: wdyt? something like this maybe should be in utils? or just manually update the args each time?
+        # from inspect import signature
+        # from typing import get_args
+        # sig = signature(self.set)
+        # expected = ", ".join([repr(a) for a in get_args(sig.parameters['value'].annotation)])
+        # raise TealInputError(
+        #     f"Got {type(value)}, expected {expected}"
+        # )
+
```

Comment:
How strongly do we feel about showing the expected types in the error message?

Ultimately, I'll follow the group's opinion.  Trying to gauge thinking.

* I'm wondering if it's sufficient to specify that the type is unsupported, and have the programmer review the method signature.
* I think the commented out introspection is neat.  I'd favor using it avoid a stale error message provided we think the formatting is clear for the programmer (example below).
  * My take is:  The message probably needs a bit of formatting to read as "expected one of these types A, B, or C".  Possibly splitting types on newlines because each type is complex.
    *  Given the work involved to format, I circle back to wondering if a vaguer message is sufficient.   

```
E       pyteal.TealInputError: Got <class 'int'>, expected typing.Sequence[pyteal.Byte], pyteal.StaticArray[pyteal.Byte, typing.Literal[32]], pyteal.ComputedValue[pyteal.StaticArray[pyteal.Byte, typing.Literal[32]]], ForwardRef('Address'), <class 'str'>, <class 'bytes'>, <class 'pyteal.Expr'>
```

Reply:
My current thoughts are that introspection is overkill for this purpose. Giving a generic error message with the type of the passed in value will be enough information for the user to debug what went wrong, since they can also very easily see the type annotations of the function args.

Reply:
^ inspect and signature on `self.get` seems too strong for this one.

We typically throw `TealInputError` with message about allowed input type, and that is assumed to be sufficient.

Reply:
👍